File: UVa 607 - Scheduling Lectures/Main.java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    static int l, c;
    static int[] topics = new int[1002];
    static int[] dp = new int[1002];
    static int[] dp2 = new int[1002];

    public static int w(int t) {
        return (t == 0) ? 0 : (t <= 10) ? -c : ((t - 10) * (t - 10));
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int tc = 1;
        while (true) {

            int n = in.nextInt();
            if (n == 0)
                break;
            if (tc != 1)
                System.out.println();
            l = in.nextInt();
            c = in.nextInt();
            for (int i = 0; i < n; i++)
                topics[i] = in.nextInt();
            dp[n] = dp2[n] = 0;
            for (int i = n - 1; i >= 0; i--) {
                dp[i] = 1 + dp[i + 1];
                int tleft = l - topics[i];
                dp2[i] = w(tleft) + dp2[i + 1];
                for (int j = i + 1; j < n && tleft >= topics[j]; j++) {
                    tleft -= topics[j];
                    if (dp[j + 1] + 1 < dp[i]
                            || (dp[j + 1] + 1 == dp[i] && dp2[j + 1] + w(tleft) < dp2[i])) {
                        dp[i] = 1 + dp[j + 1];
                        dp2[i] = dp2[j + 1] + w(tleft);
                    }
                }

            }
            System.out.printf("Case %d:\n", tc++);
            System.out.printf("Minimum number of lectures: %d\n", dp[0]);
            System.out.printf("Total dissatisfaction index: %d\n", dp2[0]);
        }
    }
}

File: UVa 607 - Scheduling Lectures/main.py
import sys
from numpypy import *
INF = 1 << 31


def solve(par):
    N, L, C, array = par
    numL = 0
    remain = 0
    for a in array:
        if remain < a:
            remain = L
            numL += 1
        remain -= a
    for i in range(1, N):
        array[i] += array[i - 1]
    dp = empty((numL, N + 1), dtype=int64)
    dp.fill(INF)

    def solve(l, start):
        '''
        start lecture l with 'start'
        '''
        if l == numL:
            if start == N:
                return 0
            else:
                return INF
        if dp[l, start] != INF:
            return dp[l, start]
        currMin = INF
        for end in range(start, N):
            if start > 0:
                s = array[end] - array[start - 1]
            else:
                s = array[end]
            if s <= L:
                v = solve(l + 1, end + 1)
                if L - s > 10:
                    v += (L - s - 10) ** 2
                elif L - s <= 10 and L - s > 0:
                    v -= C
                currMin = min(currMin, v)
            else:
                break

        dp[l, start] = currMin
        return currMin

    return 'Minimum number of lectures: %d\nTotal dissatisfaction index: %d' %\
        (numL, solve(0, 0))

if __name__ == '__main__':
    sys.stdin = open('input.txt', 'r')
    while True:
        N = int(input())
        if N == 0:
            break
        L, C = map(int, raw_input().split())
        array = []
        for i in range(N):
            array.append(int(input()))
        print(solve((N, L, C, array)))
